### PR TITLE
Refactor unregister address method

### LINKED ISSUE #149: (Refactor) change address removal method
- **What is it?** (leave one option)
    * `(Refactor)` proposal (i.e. no new features or updates, just code optimization/library updates)

https://github.com/poanetwork/poa-popa/blob/master/blockchain/contracts/ProofOfPhysicalAddress.sol#L341

In current implementation when address is removed by a user, elements of `physicalAddresses` array are shifted and then the last element is deleted:
```
[0, 1, 2, 3, 4] --> remove 2 --> [0, 1, 3, 4, 4] --> [0, 1, 3, 4]
```
Suggestion is to copy the last element of array to the place of the removed element, then delete the last element:
```
[0, 1, 2, 3, 4] --> remove 2 --> [0, 1, 4, 3, 4] --> [0, 1, 4, 3]
```
(this is for the case when removed element is not already the last one)

This method requires fewer operations and hence less gas, however the order of elements is not preserved. But it doesn't seem to be important, since they can be ordered by creation block anyway.
 
@fvictorio @matiasgaratortiz what do you think?

### LINKED ISSUE #155: Tests failing
- **What is it?** (leave one option)
    * `(Bug)` report (i.e. something doesn't work as it should)

```
Launched testrpc on port 8555
Running: ../node_modules/.bin/truffle test --network coverage
(this can take a few seconds)...
Using network 'coverage'.
Error: Error: base fee exceeds gas limit
    at runCall (/home/travis/build/poanetwork/poa-popa/blockchain/node_modules/ethereumjs-testrpc-sc/build/cli.node.js:71005:17)
    at /home/travis/build/poanetwork/poa-popa/blockchain/node_modules/ethereumjs-testrpc-sc/build/cli.node.js:12656:24
    at replenish (/home/travis/build/poanetwork/poa-popa/blockchain/node_modules/ethereumjs-testrpc-sc/build/cli.node.js:9788:17)
    at iterateeCallback (/home/travis/build/poanetwork/poa-popa/blockchain/node_modules/ethereumjs-testrpc-sc/build/cli.node.js:9773:17)
    at /home/travis/build/poanetwork/poa-popa/blockchain/node_modules/ethereumjs-testrpc-sc/build/cli.node.js:9748:16
    at /home/travis/build/poanetwork/poa-popa/blockchain/node_modules/ethereumjs-testrpc-sc/build/cli.node.js:12661:13
    at /home/travis/build/poanetwork/poa-popa/blockchain/node_modules/ethereumjs-testrpc-sc/build/cli.node.js:67133:16
    at replenish (/home/travis/build/poanetwork/poa-popa/blockchain/node_modules/ethereumjs-testrpc-sc/build/cli.node.js:67080:25)
    at /home/travis/build/poanetwork/poa-popa/blockchain/node_modules/ethereumjs-testrpc-sc/build/cli.node.js:67089:9
    at eachLimit (/home/travis/build/poanetwork/poa-popa/blockchain/node_modules/ethereumjs-testrpc-sc/build/cli.node.js:67013:36)
    at Object.InvalidResponse (/home/travis/build/poanetwork/poa-popa/blockchain/node_modules/truffle/build/webpack:/~/web3/lib/web3/errors.js:38:1)
    at /home/travis/build/poanetwork/poa-popa/blockchain/node_modules/truffle/build/webpack:/~/web3/lib/web3/requestmanager.js:86:1
    at /home/travis/build/poanetwork/poa-popa/blockchain/node_modules/truffle/build/webpack:/~/truffle-migrate/index.js:225:1
    at /home/travis/build/poanetwork/poa-popa/blockchain/node_modules/truffle/build/webpack:/~/truffle-provider/wrapper.js:134:1
    at XMLHttpRequest.request.onreadystatechange (/home/travis/build/poanetwork/poa-popa/blockchain/node_modules/truffle/build/webpack:/~/web3/lib/web3/httpprovider.js:128:1)
    at XMLHttpRequestEventTarget.dispatchEvent (/home/travis/build/poanetwork/poa-popa/blockchain/node_modules/truffle/build/webpack:/~/xhr2/lib/xhr2.js:64:1)
    at XMLHttpRequest._setReadyState (/home/travis/build/poanetwork/poa-popa/blockchain/node_modules/truffle/build/webpack:/~/xhr2/lib/xhr2.js:354:1)
    at XMLHttpRequest._onHttpResponseEnd (/home/travis/build/poanetwork/poa-popa/blockchain/node_modules/truffle/build/webpack:/~/xhr2/lib/xhr2.js:509:1)
    at IncomingMessage.<anonymous> (/home/travis/build/poanetwork/poa-popa/blockchain/node_modules/truffle/build/webpack:/~/xhr2/lib/xhr2.js:469:1)
    at emitNone (events.js:111:20)
Cleaning up...
Event trace could not be read.
Error: ENOENT: no such file or directory, open './allFiredEvents'
Exiting without generating coverage...
```

### PR DESCRIPTION
**Pull request by fvictorio**
Closes #149.
Closes #155.

Note that this will do an unnecessary assignment when the user removes the last address, but I'm not sure if adding a conditional is worth it.

I also added more tests to be sure that it's working as expected.

### CODE REVIEW

**Comment by phahulin** [comment]
I think we should add conditional for clarity.

**Comment by fvictorio** [comment]
Done. I also fixed the issue with the coverage script.

File: blockchain/contracts/ProofOfPhysicalAddress.sol [hunk 1]
```diff
@@ -342,11 +342,9 @@ contract ProofOfPhysicalAddress {
         // Remove physical address from list
         uint256 length = users[msg.sender].physicalAddresses.length;
 
-        for (uint256 i = index; i < length - 1; i++) {
-            users[msg.sender].physicalAddresses[i] = users[msg.sender].physicalAddresses[i+1];
+        if (length > 1) {
```
Comment (phahulin) [inline blockchain/contracts/ProofOfPhysicalAddress.sol]: Shouldn't this be `if (index != length -1) {` ?

Comment (fvictorio) [reply blockchain/contracts/ProofOfPhysicalAddress.sol]: Oh, that's much better. Thanks.

Comment (fvictorio) [reply blockchain/contracts/ProofOfPhysicalAddress.sol]: Done.

**Review by phahulin — APPROVED** [review]

**Pull request merged**